zipline/src/commonMain/kotlin/app/cash/zipline/Handle.kt

```diff
@@ -0,0 +1,107 @@
+import app.cash.zipline.internal.bridge.InboundBridge
+import app.cash.zipline.internal.bridge.InboundCallHandler
+import app.cash.zipline.internal.bridge.OutboundBridge
+import kotlinx.serialization.modules.SerializersModule
+import okio.Closeable
+
+abstract class Handle<T : Any> internal constructor() : Closeable {
+  fun get(serializersModule: SerializersModule): T {
+    error("unexpected call to Handle.get: is the Zipline plugin configured?")
+  }
+
+  @PublishedApi
+  internal abstract fun get(outboundBridge: OutboundBridge<T>): T
+}
+
+fun <T : Any> Handle(serializersModule: SerializersModule, service: T): Handle<T> {
+  error("unexpected call to Handle(): is the Zipline plugin configured?")
+}
+
+// Plugin-rewritten code calls the constructor of this class.
+@PublishedApi
+internal class InboundHandle<T : Any>(
+  private val inboundBridge: InboundBridge<T>
+): Handle<T>() {
+  private var name: String? = null
+  private var endpoint: Endpoint? = null
+
+  internal fun connect(endpoint: Endpoint, name: String): InboundCallHandler {
+    check(this.endpoint == null && this.name == null) { "already connected" }
+    this.name = name
+    this.endpoint = endpoint
+
+    val serializersModule = SerializersModule {
+      include(endpoint.builtInSerializersModule)
+      include(inboundBridge.serializersModule)
+    }
+
+    val context = InboundBridge.Context(serializersModule)
+    val result = inboundBridge.create(context)
+    endpoint.inboundHandlers[name] = result
+    return result
+  }
+
+  override fun get(outboundBridge: OutboundBridge<T>): T {
+    // TODO(jwilson): change generated subtypes to populate `T` on the super call.
+    throw IllegalStateException("get() on inbound handlers not yet implemented")
+  }
+
+  override fun close() {
+    val endpoint = this.endpoint
+    val name = this.name
+    if (endpoint != null && name != null) {
+      val removed = endpoint.inboundHandlers.remove(name)
+      this.endpoint = null
+      this.name = null
+      require(removed != null) { "unable to find $name: was it removed twice?" }
+    }
+  }
+}
+
+internal class OutboundHandle<T : Any> : Handle<T>() {
+  private var name: String? = null
+  private var endpoint: Endpoint? = null
+
+  internal fun connect(endpoint: Endpoint, name: String) {
+    check(this.endpoint == null && this.name == null) { "already connected" }
+    this.name = name
+    this.endpoint = endpoint
+  }
+
+  override fun get(outboundBridge: OutboundBridge<T>): T {
+    val outboundEndpoint = this.endpoint ?: throw IllegalStateException("not connected")
+    val name = this.name!!
+
+    val serializersModule = SerializersModule {
+      include(outboundEndpoint.builtInSerializersModule)
+      include(outboundBridge.serializersModule)
+    }
+
+    val context = OutboundBridge.Context(name, serializersModule, outboundEndpoint)
+    return outboundBridge.create(context)
+  }
+
+  override fun close() {
+    // TODO(jwilson): support detaching outbound handles.
```

Comment:
the only reason I haven’t done this already is it needs a bridged call to perform the removal.